### PR TITLE
Restore wallet

### LINKED ISSUE #423: Implement Restore wallet from recovery phrase
Basically reuses the Verify seed design: https://projects.invisionapp.com/d/main/default/#/console/13241316/281903495/preview

The view would ask the user to input the seed words in random order (to mitigate key loggers).

### PR DESCRIPTION
**Pull request by valentinewallace**
Closes #423. 

The storybook is not happy if you enter values and click `Next` due to the `store` values being set as if the user were initializing a new wallet, not restoring an old one. Not sure if this is OK or worth fixing?

### CODE REVIEW

File: src/action/wallet.js [hunk 1]
```diff
@@ -172,37 +185,81 @@ class WalletAction {
     await this.initWallet({
       walletPassword: password,
       seedMnemonic: this._store.seedMnemonic.toJSON(),
+      recoveryWindow: 0,
```
Comment (tanx) [inline src/action/wallet.js]: Not needed if default is set.

File: src/view/setting-unit.js [hunk 2]
```diff
@@ -28,7 +28,7 @@ const styles = StyleSheet.create({
 
 const SettingUnitView = ({ store, nav, setting }) => {
   return (
-    <Background color={color.blackDark} style={styles.wrapper}>
+    <Background color={color.blackDark}>
```
Comment (tanx) [inline src/view/setting-unit.js]: Good catch!

File: src/view/seed.js [hunk 3]
```diff
@@ -4,16 +4,21 @@ import { observer } from 'mobx-react';
...
 // Seed View
 //
 
-const SeedView = ({ store, wallet }) => (
+const SeedView = ({ store, wallet, nav }) => (
   <SplitBackground image="purple-gradient-bg" bottom={color.blackDark}>
+    <Header>
+      <BackButton onPress={() => nav.goSelectSeed()} />
+      <Button disabled onPress={() => {}} />
+    </Header>
```
Comment (tanx) [inline src/view/seed.js]: This breaks the layout...
bildschirmfoto 2018-09-04 um 14 03 30

File: src/action/wallet.js [hunk 4]
```diff
@@ -172,37 +185,81 @@ class WalletAction {
...
-    await this.unlockWallet({ walletPassword: password });
+    if (this._store.wallet.restoring) {
+      await this.initWallet({
+        walletPassword: password,
+        seedMnemonic: this._store.wallet.seedVerify.toJSON(),
+        recoveryWindow: RECOVERY_WINDOW,
+      });
+    } else {
+      await this.unlockWallet({ walletPassword: password });
+    }
```
Comment (tanx) [inline src/action/wallet.js]: Not sure I understand. Why are we using the password check screen for restoration?

File: src/computed/seed.js [hunk 5]
```diff
@@ -7,14 +7,37 @@ import { computed, extendObservable } from 'mobx';
...
+            group[0],
+            group[1],
+            group[2],
+          ];
+        }
+        return indexes;
+      } else {
+        const { seedMnemonic: words } = store;
+        return words.length ? getSeedIndexes(1, words.length, 3) : [];
+      }
```
Comment (tanx) [inline src/computed/seed.js]: Why mix seed verification logic and seed restore logic? Seems like this code would be much clean if the two were separated.

File: src/computed/seed.js [hunk 6]
```diff
@@ -7,14 +7,37 @@ import { computed, extendObservable } from 'mobx';
...
       const { seedVerifyIndexes } = store;
-      const c0 = formatOrdinal(seedVerifyIndexes[0]);
-      const c1 = formatOrdinal(seedVerifyIndexes[1]);
-      const c2 = formatOrdinal(seedVerifyIndexes[2]);
+      const startingIndex = store.wallet.restoring
+        ? store.wallet.restoreIndex
+        : 0;
+      const c0 = formatOrdinal(seedVerifyIndexes[startingIndex]);
+      const c1 = formatOrdinal(seedVerifyIndexes[startingIndex + 1]);
+      const c2 = formatOrdinal(seedVerifyIndexes[startingIndex + 2]);
```
Comment (tanx) [inline src/computed/seed.js]: Same point as above.

File: src/view/restore-wallet.js [hunk 7]
```diff
@@ -44,18 +44,22 @@ const RestoreWalletView = ({ store, nav, wallet }) => (
...
-            seedIndex={seedIndex}
-            value={store.wallet.seedVerify[i]}
-            onChangeText={word => wallet.setSeedVerify({ word, index: i })}
-            key={i}
-            autoFocus={i === 0}
-            onSubmitEditing={() => wallet.checkSeed()}
-          />
-        ))}
+        {store.seedVerifyIndexes
+          .slice(store.wallet.restoreIndex, store.wallet.restoreIndex + 3)
```
Comment (tanx) [inline src/view/restore-wallet.js]: Might be cleaner to pull out the `slice(...)` logic into a computed attribute?

File: src/view/select-seed.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  </Background>
+);
+
+SelectSeedView.propTypes = {
+  store: PropTypes.object.isRequired,
+  wallet: PropTypes.object.isRequired,
+  nav: PropTypes.object.isRequired,
+};
+
+export default observer(SelectSeedView);
```
Comment (tanx) [inline src/view/select-seed.js]: I cleaned this view up a bit.

File: src/action/wallet.js [hunk 9]
```diff
@@ -181,18 +204,48 @@ class WalletAction {
...
+        recovery_window: recoveryWindow,
       });
       this._store.walletUnlocked = true;
       this._nav.goSeedSuccess();
     } catch (err) {
-      this._notification.display({ msg: 'Initializing wallet failed', err });
+      this._notification.display({
+        type: 'error',
+        msg: `Initializing wallet failed: ${JSON.stringify(err)}`,
+      });
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply src/action/wallet.js]: When I leave it the way it was, the error appears like this on the screen:
deepinscreenshot_select-area_20180905130929

Which allows the user to click "Show error logs," bringing them to the CLI, and from there they can navigate to the home screen even though their wallet hasn't been initialized. 

I do think it's a good idea to give some level of detail in the notification's error message.. Let me know if you like the new solution, I just show the `detalis` field of the error.

File: src/computed/seed.js [hunk 10]
```diff
@@ -17,6 +17,37 @@ const ComputedSeed = store => {
       const c2 = formatOrdinal(seedVerifyIndexes[2]);
       return `Type the ${c0}, ${c1}, and ${c2} words of your recovery phrase.`;
     }),
+    restoreIndexes: computed(() => {
+      let indexes = [...Array(24).keys()].map(x => ++x);
+      // Shuffle the array.
+      for (let i = 23; i > 0; i--) {
+        const j = Math.floor(Math.random() * (i + 1));
+        [indexes[i], indexes[j]] = [indexes[j], indexes[i]];
+      }
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply src/computed/seed.js]: Ah, I thought we were trying to mitigate keyloggers. Fixed!

File: src/view/restore-password.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        <FormSubText>Please enter your password.</FormSubText>
+        <FormStretcher>
+          <InputField
+            style={styles.input}
+            placeholder="Password"
+            secureTextEntry={true}
+            autoFocus={true}
+            value={store.wallet.password}
+            onChangeText={password => wallet.setPassword({ password })}
+            onSubmitEditing={() => wallet.checkPassword()}
```
[the remark this answers is not in the document]
Comment (valentinewallace) [reply src/view/restore-password.js]: oo.. thanks! fixed

File: src/action/wallet.js [hunk 12]
```diff
@@ -181,18 +204,48 @@ class WalletAction {
...
+        recovery_window: recoveryWindow,
       });
       this._store.walletUnlocked = true;
       this._nav.goSeedSuccess();
     } catch (err) {
-      this._notification.display({ msg: 'Initializing wallet failed', err });
+      this._notification.display({
+        type: 'error',
+        msg: `Initializing wallet failed: ${err.details}`,
+      });
```
Comment (tanx) [inline src/action/wallet.js]: The above does the same in one line. Why the change?

Comment (valentinewallace) [reply src/action/wallet.js]: Sorry, did you see this comment? https://github.com/lightninglabs/lightning-app/pull/612#discussion_r215411823 This is the reason I changed it, you may not be getting the same issue?
